optuna/samplers/_grid.py

```diff
@@ -251,6 +252,12 @@ def _get_unvisited_grid_ids(self, study: Study) -> List[int]:
 
         return list(unvisited_grids)
 
+    @staticmethod
+    def _grid_value_equal(value1: GridValueType, value2: GridValueType) -> bool:
+        value1_is_nan = isinstance(value1, Real) and np.isnan(float(value1))
+        value2_is_nan = isinstance(value2, Real) and np.isnan(float(value2))
+        return (value1 == value2) or (value1_is_nan and value2_is_nan)
```

Comment:
I thought making `_grid_value_equal` a function instead of the static method might be more consistent with #3567, but this is not a strong opinion and I leave it to you.

Reply:
I think it would be nicer to be a static method, because this method is never used from outside the class.

Reply:
Sure, then I'll merge this PR.